File: frontend/app/api/line.ts
import apiFetch from "./api";

export type Line = {
  INV_NUMBER: number;
  LINE_NUMBER: number;
  PROD_CODE: string | null;
  LINE_UNITS: number | null;
  LINE_PRICE: number | null;
  LINE_AMOUNT: number | null;
};

export type LineUpdate = {
  PROD_CODE: string;
  LINE_UNITS: number;
  LINE_PRICE: number;
  LINE_AMOUNT: number;
};

export async function saveLine(newLine: Line) {
  const response = await apiFetch("/lines/save", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(newLine),
  });

  if (!response) {
    throw new Error("Failed to save line.");
  }

  return response.ok;
}

export async function updateLine(invoiceId: number, lineId: number, updatedLine: LineUpdate) {
  const response = await apiFetch(`/lines/update/${invoiceId}/${lineId}`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(updatedLine),
  });

  if (!response) {
    throw new Error("Failed to update line.");
  }

  return response.ok;
}

export async function deleteLineById(invoiceId: number, lineId: number) {
  const response = await apiFetch(`/lines/delete/${invoiceId}/${lineId}`, {
    method: "DELETE",
  });

  if (!response) {
    throw new Error("Failed to delete line.");
  }

  return response.ok;
}

export async function findAllLines() {
  const response = await apiFetch("/lines/all");

  if (!response) {
    throw new Error("Failed to retrieve lines.");
  }

  const data = (await response.json()) as Line[];
  return data;
}

File: backend/app/api/routes/line.py
from fastapi import APIRouter, Depends, HTTPException

from app.database import get_db
from app.models import LineCreate, LineUpdate

router = APIRouter(prefix="/lines")


@router.get("/all")
def get_all_lines(db=Depends(get_db)):
    cursor = db.cursor(dictionary=True)

    query = "SELECT * FROM LINE"
    cursor.execute(query)
    return cursor.fetchall()


@router.post("/save")
def save_line(line: LineCreate, db=Depends(get_db)):
    line_data = (
        line.INV_NUMBER,
        line.LINE_NUMBER,
        line.PROD_CODE,
        line.LINE_UNITS,
        line.LINE_PRICE,
        line.LINE_AMOUNT,
    )

    try:
        cursor = db.cursor()
        cursor.execute(
            """
            INSERT INTO LINE
            (INV_NUMBER, LINE_NUMBER, PROD_CODE, LINE_UNITS, LINE_PRICE, LINE_AMOUNT)
            VALUES (%s, %s, %s, %s, %s, %s)
            """,
            line_data,
        )

        db.commit()
    except:
        raise HTTPException(status_code=500, detail="Failed to save line")


@router.put("/update/{invoice_id}/{line_id}")
def update_line(invoice_id: int, line_id: int, line: LineUpdate, db=Depends(get_db)):
    line_data = (
        line.PROD_CODE,
        line.LINE_UNITS,
        line.LINE_PRICE,
        line.LINE_AMOUNT,
    )
    try:
        cursor = db.cursor(dictionary=True)
        cursor.execute(
            """
            UPDATE LINE
            SET PROD_CODE = %s, LINE_UNITS = %s, LINE_PRICE = %s, LINE_AMOUNT = %s
            WHERE INV_NUMBER = %s AND LINE_NUMBER = %s
            """,
            (*line_data, invoice_id, line_id),
        )
        db.commit()
    except:
        raise HTTPException(status_code=500, detail="Failed to update line")


@router.delete("/delete/{invoice_id}/{line_id}", status_code=204)
def delete_line_by_id(invoice_id: int, line_id: int, db=Depends(get_db)):
    cursor = db.cursor()

    cursor.execute(
        """
        DELETE FROM LINE WHERE INV_NUMBER = %s AND LINE_NUMBER = %s""",
        [invoice_id, line_id],
    )

    db.commit()
